### PR TITLE
Corrections to #525 changes to BOOKING ARRANGEMENTs and SERVCIE BOOKING ARRANGEMENTS

### PR DESCRIPTION
**Pull request by nick-knowles**
Correct the [#525](https://github.com/NeTEx-CEN/NeTEx/issues/525) implementation of BOOKING ARRANGEMENT and SERVICE BOOKING ARRANGEMENT. 

There were a number of issues with the XML encoding, including

SERVICE BOOKING ARRANGEMENT should be a reusable XML element with its own _ServiceBooking_VersionStructure_ element  (and not just be declared inline) and **_ServiceBookingRef_** element)   

The _ServiceBooking_VersionStructure_ should inherit from _Booking_VersionStructure_

SERVICE BOOKING ARRANGEMENT should have an uniqueness constraint. 

The  **_revisedbookingArrangements_** should  be refactored upto LINE from FLEXIBLE LINE (and not be added to FLEXIBLE LINE)  Elsewhere should be consistent about using a choice for new way vs deprecated. Also suggest placing the preferred new way first and the deprecated method second. Also document which attributes are new in v2.0

ALso  examples neede revising  to test new encoding 

Note that for now I have not made **_BookingElement_** substitutable by **_ServiceBookingElement_** . ISince the current use of **_ServiceBookingElement_**  is limited).

### CODE REVIEW

**Comment by nick-knowles** [comment]
This is the UML summarising the changes -  (Over fifty other diagrams are updated )

NeTEx 
image

Transmodel
image

**Comment by skinkie** [comment]
@nick-knowles we cannot work with this pull request.

File: xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd [hunk 1]
```diff
@@ -161,9 +210,9 @@ Rail transport, Roads and Road transport
 			</xsd:complexContent>
 		</xsd:complexType>
 	</xsd:element>
-	<xsd:complexType name="BookingArrangement_VersionStructure" abstract="true">
+	<xsd:complexType name="BookingArrangement_VersionStructure" abstract="false">
 		<xsd:annotation>
-			<xsd:documentation>Type for a BookingArrangement</xsd:documentation>
+			<xsd:documentation>Type for a OOKING ARRANGEMENT</xsd:documentation>
```
Comment (ue71603) [inline xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd:215]:
```suggestion
			<xsd:documentation>Type for a BOOKING ARRANGEMENT</xsd:documentation>
```

File: xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd [hunk 2]
```diff
@@ -234,107 +283,179 @@ Rail transport, Roads and Road transport
...
 				<xsd:annotation>
-					<xsd:documentation>Minimum period for which a booking can be mad +v.1.2.2</xsd:documentation>
+					<xsd:documentation>Contact for Booking. +v1.1</xsd:documentation>
 				</xsd:annotation>
 			</xsd:element>
-			<xsd:element name="MaximumBookingDuration" type="xsd:duration" minOccurs="0">
+			<xsd:element name="BookingMethods" type="BookingMethodListOfEnumerations" minOccurs="0">
 				<xsd:annotation>
-					<xsd:documentation>Maximum period for which a booking can be mad +v.1.2.2</xsd:documentation>
+					<xsd:documentation>Allowed Ways of Making a BOOKING.</xsd:documentation>
```
Comment (ue71603) [inline xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd:307]:
```suggestion
					<xsd:documentation>Allowed ways of making a BOOKING.</xsd:documentation>
```

File: xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd [hunk 3]
```diff
@@ -234,107 +283,179 @@ Rail transport, Roads and Road transport
...
 				<xsd:annotation>
-					<xsd:documentation>Whether a deposit is required. +v1.2.2</xsd:documentation>
+					<xsd:documentation>Who can make a booking. Default is public.</xsd:documentation>
 				</xsd:annotation>
 			</xsd:element>
-			<xsd:element name="BookingChargeType" type="BookingChargeTypeEnumeration" minOccurs="0">
+			<xsd:element name="BookWhen" type="PurchaseWhenEnumeration" minOccurs="0">
 				<xsd:annotation>
-					<xsd:documentation>Nature of booking fee. v1.2.2</xsd:documentation>
+					<xsd:documentation>When Booking can be made. +V1.1</xsd:documentation>
```
Comment (ue71603) [inline xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd:317]:
```suggestion
					<xsd:documentation>When a booking can be made. +V1.1</xsd:documentation>
```

File: xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd [hunk 4]
```diff
@@ -234,107 +283,179 @@ Rail transport, Roads and Road transport
...
+				</xsd:annotation>
+			</xsd:element>
+			<xsd:element name="BuyWhen" type="PurchaseMomentListOfEnumerations" minOccurs="0">
+				<xsd:annotation>
+					<xsd:documentation>When purchase can be made.  +V1.1</xsd:documentation>
```
Comment (ue71603) [inline xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd:322]:
```suggestion
					<xsd:documentation>When purchase can be made. +V1.1</xsd:documentation>
```

File: xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd [hunk 5]
```diff
@@ -234,107 +283,179 @@ Rail transport, Roads and Road transport
...
+				</xsd:annotation>
+			</xsd:element>
+			<xsd:element name="MinimumBookingPeriod" type="xsd:duration" minOccurs="0">
+				<xsd:annotation>
+					<xsd:documentation>Minimum interval in advance of departure day or time that Service may be ordered.</xsd:documentation>
+				</xsd:annotation>
+			</xsd:element>
+			<xsd:element name="MaximumBookingPeriod" type="xsd:duration" minOccurs="0">
+				<xsd:annotation>
+					<xsd:documentation>Maximum interval in advance of departure day or time that Service may be ordered. +V1.2..2</xsd:documentation>
```
Comment (ue71603) [inline xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd:337]:
```suggestion
					<xsd:documentation>Maximum interval in advance of departure day or time that a service may be ordered. +V1.2..2</xsd:documentation>
```

File: xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd [hunk 6]
```diff
@@ -234,107 +283,179 @@ Rail transport, Roads and Road transport
...
-		</xsd:annotation>
-		<xsd:sequence>
-			<xsd:group ref="BookingArrangementGroup"/>
-		</xsd:sequence>
-	</xsd:complexType>
-	<xsd:complexType name="ServiceBookingArrangementsStructure" abstract="false">
+	<xsd:group name="ServiceBookingArrangementGroup">
 		<xsd:annotation>
-			<xsd:documentation>Type for SERVICE BOOKING ARRANGEMENTs.</xsd:documentation>
+			<xsd:documentation>Elements for A SERVICE BOOKING ARRANGEMENT.</xsd:documentation>
```
Comment (ue71603) [inline xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd:434]:
```suggestion
			<xsd:documentation>Elements for a SERVICE BOOKING ARRANGEMENT.</xsd:documentation>
```

File: xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd [hunk 7]
```diff
@@ -234,107 +283,179 @@ Rail transport, Roads and Road transport
...
-			<xsd:element name="ServiceBookingArrangement" maxOccurs="unbounded">
-				<xsd:complexType>
-					<xsd:sequence>
-						<xsd:element name="BookingArrangement" type="BookingArrangementRef_RelStructure"/>
-						<xsd:group ref="ServiceBookingArrangementGroup"/>
-					</xsd:sequence>
-				</xsd:complexType>
+			<xsd:element name="MinimumBookingDuration" type="xsd:duration" minOccurs="0">
+				<xsd:annotation>
+					<xsd:documentation>Minimum period for which a booking can be mad +v.1.2.2</xsd:documentation>
```
Comment (ue71603) [inline xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd:439]:
```suggestion
					<xsd:documentation>Minimum period for which a booking can be made. +v.1.2.2</xsd:documentation>
```

File: xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd [hunk 8]
```diff
@@ -234,107 +283,179 @@ Rail transport, Roads and Road transport
...
+				</xsd:annotation>
+			</xsd:element>
+			<xsd:element name="MaximumBookingDuration" type="xsd:duration" minOccurs="0">
+				<xsd:annotation>
+					<xsd:documentation>Maximum period for which a booking can be mad +v.1.2.2</xsd:documentation>
```
Comment (ue71603) [inline xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd:444]:
```suggestion
					<xsd:documentation>Maximum period for which a booking can be made. +v.1.2.2</xsd:documentation>
```

File: xsd/netex_part_1/part1_ifopt/netex_ifopt_path_support.xsd [hunk 9]
```diff
@@ -319,6 +319,45 @@ Rail transport, Roads and Road transport
 			<xsd:enumeration value="other"/>
 		</xsd:restriction>
 	</xsd:simpleType>
+	<xsd:simpleType name="FlooringStatusEnumeration">
+		<xsd:annotation>
+			<xsd:documentation>Allowed values for the status of the flooring</xsd:documentation>
```
Comment (ue71603) [inline xsd/netex_part_1/part1_ifopt/netex_ifopt_path_support.xsd:324]:
```suggestion
			<xsd:documentation>Allowed values for the status of the flooring.</xsd:documentation>
```

File: xsd/netex_part_1/part1_ifopt/netex_ifopt_path_support.xsd [hunk 10]
```diff
@@ -319,6 +319,45 @@ Rail transport, Roads and Road transport
...
+			<xsd:enumeration value="good"/>
+			<xsd:enumeration value="worn"/>
+			<xsd:enumeration value="discomfortable"/>
+			<xsd:enumeration value="hazardous"/>
+			<xsd:enumeration value="unknown"/>
+		</xsd:restriction>
+	</xsd:simpleType>
+	<xsd:simpleType name="TactileGuidingStripStatusEnumeration">
+		<xsd:annotation>
+			<xsd:documentation>Allowed values for the status of the Tactile Guiding Strip</xsd:documentation>
```
Comment (ue71603) [inline xsd/netex_part_1/part1_ifopt/netex_ifopt_path_support.xsd:336]:
```suggestion
			<xsd:documentation>Allowed values for the status of the Tactile Guiding Strip.</xsd:documentation>
```

**Review by ue71603 — APPROVED** [review]

File: examples/standards/norway/vehicles/submodel-VehicleType.xml [hunk 11]
```diff
@@ -4,7 +4,7 @@
 	<Description>V</Description>
 	<SelfPropelled>true</SelfPropelled>
 	<TypeOfFuel>diesel</TypeOfFuel>
-	<EuroClass>Euro 6</EuroClass>
+	<!-- <EuroClass>Euro 6</EuroClass> -->
```
Comment (skinkie) [inline examples/standards/norway/vehicles/submodel-VehicleType.xml:7]: Should we comment out code in examples?

File: examples/functions/newModes/NewModes-CarPoolingExample.xml [hunk 12]
```diff
@@ -445,6 +445,23 @@ In addition CUSTOMER PURCHASE PACKAGES showing the parameters that might be held
...
+									<BookingContact>
+										<Url>https://www.ryde.eu/booking</Url>
+									</BookingContact>
+									<BookingMethods>mobileApp online</BookingMethods>
+									<BookingAccess>authorisedPublic</BookingAccess>
+									<BookWhen>advanceAndDayOfTravel</BookWhen>
+									<BuyWhen>onCheckOut</BuyWhen>
+									<MinimumBookingPeriod>PT2H</MinimumBookingPeriod>
+									<BookingUrl>https://www.ryde.eu/booking</BookingUrl>
+									<BookingNote>Booklings made using app. Bookings can be made up to two hour before travel. </BookingNote>
```
Comment (skinkie) [inline examples/functions/newModes/NewModes-CarPoolingExample.xml:459]:
```suggestion
									<BookingNote>Booklings made using app. Bookings can be made up to two hour before travel.</BookingNote>
```

File: xsd/netex_part_1/part1_ifopt/netex_ifopt_path_support.xsd [hunk 13]
```diff
@@ -319,6 +319,45 @@ Rail transport, Roads and Road transport
...
+		</xsd:restriction>
+	</xsd:simpleType>
+	<xsd:simpleType name="TactileGuidingStripStatusEnumeration">
+		<xsd:annotation>
+			<xsd:documentation>Allowed values for the status of the Tactile Guiding Strip.</xsd:documentation>
+		</xsd:annotation>
+		<xsd:restriction base="xsd:NMTOKEN">
+			<xsd:enumeration value="goodAndContrasted">
+				<xsd:annotation>
+					<xsd:documentation>when there guiding strip has the expected quality level</xsd:documentation>
```
Comment (skinkie) [inline xsd/netex_part_1/part1_ifopt/netex_ifopt_path_support.xsd:341]:
```suggestion
					<xsd:documentation>when there guiding strip has the expected quality level.</xsd:documentation>
```

File: xsd/netex_part_1/part1_ifopt/netex_ifopt_path_support.xsd [hunk 14]
```diff
@@ -319,6 +319,45 @@ Rail transport, Roads and Road transport
...
+				</xsd:annotation>
+			</xsd:enumeration>
+			<xsd:enumeration value="primitive">
+				<xsd:annotation>
+					<xsd:documentation>when there is no guide strip as such but a decorative element or ledge can be used for this purpose</xsd:documentation>
```
Comment (skinkie) [inline xsd/netex_part_1/part1_ifopt/netex_ifopt_path_support.xsd:346]:
```suggestion
					<xsd:documentation>when there is no guide strip as such but a decorative element or ledge can be used for this purpose.</xsd:documentation>
```

File: xsd/netex_part_1/part1_ifopt/netex_ifopt_path_support.xsd [hunk 15]
```diff
@@ -319,6 +319,45 @@ Rail transport, Roads and Road transport
...
+				</xsd:annotation>
+			</xsd:enumeration>
+			<xsd:enumeration value="incorrect">
+				<xsd:annotation>
+					<xsd:documentation>when there is a guide strip but it does not correspond to a path that can actually be used (passes through the middle of an obstacle without going around it for example)</xsd:documentation>
```
Comment (skinkie) [inline xsd/netex_part_1/part1_ifopt/netex_ifopt_path_support.xsd:351]:
```suggestion
					<xsd:documentation>When there is a guide strip but it does not correspond to a path that can actually be used (passes through the middle of an obstacle without going around it for example).</xsd:documentation>
```

File: xsd/netex_part_1/part1_ifopt/netex_ifopt_path_support.xsd [hunk 16]
```diff
@@ -319,6 +319,45 @@ Rail transport, Roads and Road transport
...
+				</xsd:annotation>
+			</xsd:enumeration>
+			<xsd:enumeration value="bad">
+				<xsd:annotation>
+					<xsd:documentation>when the guide strip is degraded and difficult to interpret</xsd:documentation>
```
Comment (skinkie) [inline xsd/netex_part_1/part1_ifopt/netex_ifopt_path_support.xsd:356]:
```suggestion
					<xsd:documentation>When the guide strip is degraded and difficult to interpret.</xsd:documentation>
```

**Comment by skinkie** [comment]
There are pavement related changes here. Please be aware with merging. This will might create serious conflicts.

File: xsd/netex_part_1/part1_ifopt/netex_ifopt_path_version.xsd [hunk 17]
```diff
@@ -186,11 +191,21 @@ Rail transport, Roads and Road transport
 					<xsd:documentation>Nature of the  tactile warning strips (in the direction of the PATH LINK). +v1.1</xsd:documentation>
 				</xsd:annotation>
 			</xsd:element>
+			<xsd:element name="TactileWarningStripContrast" type="xsd:boolean" minOccurs="0" default="true">
+				<xsd:annotation>
+					<xsd:documentation>indicates whether the color of the possible warning strip is sufficiently contrasted with the color of the floor. To compare with the OpenStreetMap tag tactile_paving https://wiki.openstreetmap.org/wiki/FR:Key:tactile_paving</xsd:documentation>
```
Comment (trurlurl) [inline xsd/netex_part_1/part1_ifopt/netex_ifopt_path_version.xsd:196]:
```suggestion
					<xsd:documentation>Indicates whether the color of the possible warning strip is sufficiently contrasted with the color of the floor. To compare with the OpenStreetMap tag tactile_paving https://wiki.openstreetmap.org/wiki/FR:Key:tactile_paving</xsd:documentation>
```

File: xsd/netex_part_1/part1_ifopt/netex_ifopt_path_version.xsd [hunk 18]
```diff
@@ -186,11 +191,21 @@ Rail transport, Roads and Road transport
...
+				</xsd:annotation>
+			</xsd:element>
 			<xsd:element name="TactileGuidingStrip" type="xsd:boolean" minOccurs="0">
 				<xsd:annotation>
 					<xsd:documentation>Indicates whether there are guiding strips. +v1.1</xsd:documentation>
 				</xsd:annotation>
 			</xsd:element>
+			<xsd:element name="TactileGuidingStripStatus" type="TactileGuidingStripStatusEnumeration" minOccurs="0" default="goodAndContrasted">
+				<xsd:annotation>
+					<xsd:documentation>Provides additional details on possible guide strip.  To compare with the OpenStreetMap tag tactile_paving https://wiki.openstreetmap.org/wiki/FR:Key:tactile_paving</xsd:documentation>
```
Comment (trurlurl) [inline xsd/netex_part_1/part1_ifopt/netex_ifopt_path_version.xsd:206]:
```suggestion
					<xsd:documentation>Provides additional details on possible guiding strip. To compare with the OpenStreetMap tag tactile_paving https://wiki.openstreetmap.org/wiki/FR:Key:tactile_paving</xsd:documentation>
```

File: xsd/netex_part_1/part1_networkDescription/netex_flexibleNetwork_version.xsd [hunk 19]
```diff
@@ -104,14 +106,9 @@ Rail transport, Roads and Road transport
 			</xsd:element>
 			<xsd:choice>
 				<xsd:annotation>
-					<xsd:documentation>BookingArrangementGroup direct use is DEPRECATED: please use bookingArrangements</xsd:documentation>
+					<xsd:documentation>BookingArrangementGroup direct use is DEPRECATED: please use bookingArrangements on LINE SUpoertype</xsd:documentation>
```
Comment (trurlurl) [inline xsd/netex_part_1/part1_networkDescription/netex_flexibleNetwork_version.xsd:109]:
```suggestion
					<xsd:documentation>BookingArrangementGroup direct use is DEPRECATED: please use bookingArrangements on LINE Supertype.</xsd:documentation>
```

File: xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd [hunk 20]
```diff
@@ -234,107 +283,179 @@ Rail transport, Roads and Road transport
...
+					</xsd:sequence>
+					<xsd:attribute name="id" type="BookingArrangementIdType" use="required"/>
+				</xsd:restriction>
+			</xsd:complexContent>
+		</xsd:complexType>
+	</xsd:element>
+	<xsd:complexType name="ServiceBookingArrangement_VersionStructure" abstract="false">
 		<xsd:annotation>
-			<xsd:documentation>Type for containment in frame of BOOKING ARRANGEMENTs</xsd:documentation>
+			<xsd:documentation>Type for a SERVICE BOOKING ARRANGEMENT</xsd:documentation>
```
Comment (trurlurl) [inline xsd/netex_framework/netex_reusableComponents/netex_serviceRestrictions_version.xsd:418]:
```suggestion
			<xsd:documentation>Version of a SERVICE BOOKING ARRANGEMENT.</xsd:documentation>
```

File: xsd/netex_part_1/part1_ifopt/netex_assistanceBooking_version.xsd [hunk 21]
```diff
@@ -139,16 +141,18 @@
...
-			</xsd:element>
-			<xsd:element name="bookingArrangements" type="bookingArrangements_RelStructure" minOccurs="0">
-				<xsd:annotation>
-					<xsd:documentation>Set of possible Booking Arrangements for this booking service</xsd:documentation>
-				</xsd:annotation>
-			</xsd:element>
+			<xsd:choice>
+				<xsd:element name="bookingArrangements" type="bookingArrangements_RelStructure" minOccurs="0">
+					<xsd:annotation>
+						<xsd:documentation>Set of possible Booking Arrangements for this booking servic +v2.0</xsd:documentation>
```
Comment (trurlurl) [inline xsd/netex_part_1/part1_ifopt/netex_assistanceBooking_version.xsd:147]:
```suggestion
						<xsd:documentation>Set of possible Booking Arrangements for this booking service. +v2.0</xsd:documentation>
```

File: xsd/netex_part_3/part3_fares/netex_usageParameterBooking_version.xsd [hunk 22]
```diff
@@ -388,16 +392,18 @@ Rail transport, Roads and Road transport
...
-			</xsd:element>
-			<xsd:element name="bookingArrangements" type="bookingArrangements_RelStructure" minOccurs="0">
-				<xsd:annotation>
-					<xsd:documentation>Set of possible Booking Arrangements for this USAGE PARAMETER. </xsd:documentation>
-				</xsd:annotation>
-			</xsd:element>
+			<xsd:choice>
+				<xsd:element name="bookingArrangements" type="bookingArrangements_RelStructure" minOccurs="0">
+					<xsd:annotation>
+						<xsd:documentation>Set of possible Booking Arrangements for Cancellations  +v2.0</xsd:documentation>
```
Comment (trurlurl) [inline xsd/netex_part_3/part3_fares/netex_usageParameterBooking_version.xsd:398]:
```suggestion
						<xsd:documentation>Set of possible Booking Arrangements for Cancellations.  +v2.0</xsd:documentation>
```

File: examples/standards/norway/vehicles/submodel-VehicleType.xml (hunk 11 above)
Comment (ue71603) [reply examples/standards/norway/vehicles/submodel-VehicleType.xml:7]: sometimes, but not here when it is without an explanation. I removed it.

**Review by trurlurl — APPROVED** [review]

**Comment by Aurige** [comment]
This PR also integrates changes from https://github.com/NeTEx-CEN/NeTEx/pull/624 (Flooringstatus, etc.) ! they should not be mixed

**Comment by ue71603** [comment]
> This PR also integrates changes from #624 (Flooringstatus, etc.) ! they should not be mixed

@Aurige  @skinkie  #624 will be merged into EPIAP. and then #525?  Probably it will cause a merge conflict and we can fix it then?

**Comment by ue71603** [comment]
@Aurige  fixed the conflicts. Is this mergable in your view?

**Review by skinkie — APPROVED** [review]

**Review by Aurige — APPROVED** [review]

**Comment by Aurige** [comment]
Parts 1,2,3 and 5 updated

**Pull request merged**